llama-index-core/llama_index/core/agent/introspective/step.py

```diff
@@ -0,0 +1,312 @@
+        llm: Optional[FunctionCallingLLM] = None,
+        verbose: bool = False,
+        max_function_calls: int = DEFAULT_MAX_FUNCTION_CALLS,
+        callback_manager: Optional[CallbackManager] = None,
+        system_prompt: Optional[str] = None,
+        prefix_messages: Optional[List[ChatMessage]] = None,
+        **kwargs: Any,
+    ) -> "IntrospectiveAgentWorker":
+        """Create an IntrospectiveAgentWorker from a list of tools.
+
+        Similar to `from_defaults` in other classes, this method will
+        infer defaults for a variety of parameters, including the LLM,
+        if they are not specified.
+
+        """
+        tools = tools or []
+
+        llm = llm or Settings.llm
+        if callback_manager is not None:
+            llm.callback_manager = callback_manager
+
+        if system_prompt is not None:
+            if prefix_messages is not None:
+                raise ValueError(
+                    "Cannot specify both system_prompt and prefix_messages"
+                )
+            prefix_messages = [ChatMessage(content=system_prompt, role="system")]
+
+        prefix_messages = prefix_messages or []
+
+        return cls(
+            tools=tools,
+            tool_retriever=tool_retriever,
+            main_agent_worker=main_agent_worker,
+            reflective_agent_worker=reflective_agent_worker,
+            llm=llm,
+            prefix_messages=prefix_messages,
+            verbose=verbose,
+            max_function_calls=max_function_calls,
+            callback_manager=callback_manager,
+            **kwargs,
+        )
+
+    def initialize_step(self, task: Task, **kwargs: Any) -> TaskStep:
+        """Initialize step from task."""
+        # temporary memory for new messages
+        main_memory = ChatMemoryBuffer.from_defaults()
+        reflective_memory = ChatMemoryBuffer.from_defaults()
+
+        # initialize task state
+        task_state = {
+            "main": {
+                "memory": main_memory,
+                "sources": [],
+            },
+            "reflection": {"memory": reflective_memory, "sources": []},
+        }
+        task.extra_state.update(task_state)
+
+        return TaskStep(
+            task_id=task.task_id,
+            step_id=str(uuid.uuid4()),
+            input=task.input,
+        )
+
+    def get_tools(self, input: str) -> List[AsyncBaseTool]:
+        """Get tools."""
+        return [adapt_to_async_tool(t) for t in self._get_tools(input)]
+
+    def get_all_messages(self, task: Task) -> List[ChatMessage]:
+        return (
+            self.prefix_messages
+            + task.memory.get()
+            + task.extra_state["main"]["memory"].get_all()
+            + task.extra_state["reflection"]["memory"].get_all()
+        )
+
+    @trace_method("run_step")
+    def run_step(self, step: TaskStep, task: Task, **kwargs: Any) -> TaskStepOutput:
+        """Run step."""
+        # run main agent
+        main_agent = self._main_agent_worker.as_agent()
```

Comment:
by calling `as_agent()` every time, we are starting with a fresh memory -- not sure if intentional or not

Reply:
Right, unless I pass in `chat_history` or `memory` as part of `kwargs`?